lib/rdkafka/producer.rb

```diff
@@ -9,11 +11,12 @@ class Producer
 
     # @private
     def initialize(native_kafka)
+      @id = SecureRandom.uuid
       @closing = false
       @native_kafka = native_kafka
 
       # Makes sure, that the producer gets closed before it gets GCed by Ruby
-      ObjectSpace.define_finalizer(self, proc { close })
+      ObjectSpace.define_finalizer(@id, proc { close })
```

Comment:
I don't quite get your point to be honest.

If you are pointing out, that it won't work upon GC.start that is true and it is something I am aware.

The goal of this update is to remove warnings not alter the behaviour. The behaviour here (that I introduced) is to finalize the producer upon Ruby VM shutdown to prevent VM crashes, not to handle all of the cases in which the producer is disposed. 

> I guess we could have a reference to the instance within the Proc as of 2.7.0 as per the following code:

As far as I remember, the way closures work in Ruby injects the `self` anyhow, that's why I'm able to run `#close` without having to pass the reference. And this is how it operates, what changes here is the fact that we do not reference the producer **within** the finalizers registry by eliminating it being passed as the first argument.

Reply:
@mensfeld I was quoting https://www.mikeperham.com/2010/02/24/the-trouble-with-ruby-finalizers/ that was mentioned in the issue which suggests using a class method to define the finalizer proc where self does not evaluate to be the instance of the class.

I think that your solution not only eliminates the warning but also closes any producers by invoking the finalizer after they are GC’d and is demonstrated by the code snippet. This snippet is a variation of https://github.com/appsignal/rdkafka-ruby/pull/115#issuecomment-622556449.

I could’ve been more clearer that I do not have any concerns with your change and only find it interesting :)

Reply:
@Adithya-copart Mikes case is a bit different. Here the purpose is to GC before the VM shutdown. Mike points out that this type of reference prevents the cleanup when GC kicks in as GC will hold the reference via the closure. But that is a different case. Here we are interested mostly in finalizing the producer upon VM shutdown not to cause a crash.

Nonetheless happy you enjoy it :)

P.S. @thijsc good you are back ;)

Reply:
Yeah sorry I was deep into working on some other things :-)